api/handlers.go

```diff
@@ -557,6 +564,122 @@ func (si *ServerImplementation) LookupApplicationByID(ctx echo.Context, applicat
+		return badRequest(ctx, err.Error())
+	}
+
+	q := idb.ApplicationBoxQuery{
+		ApplicationID: applicationID,
+		BoxName:       boxName,
+	}
+	appid, boxes, round, err := si.fetchApplicationBoxes(ctx.Request().Context(), q)
+
+	if err != nil {
+		msg := fmt.Sprintf("%s: round=?=%d, appid=%d, boxName=%s", errFailedLookingUpBoxes, round, applicationID, encodedBoxName)
+		return indexerError(ctx, fmt.Errorf("%s: %w", msg, err))
+	}
+
+	if len(boxes) == 0 {
+		return notFound(ctx, fmt.Sprintf("%s: round=%d, appid=%d, boxName=%+v", errNoApplicationsFound, round, applicationID, encodedBoxName))
+	}
+
+	if appid != generated.ApplicationId(applicationID) {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, wrong appid=%d, boxName=%s", errWrongAppidFound, round, applicationID, appid, encodedBoxName))
+	}
+
+	if len(boxes) > 1 {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, boxName=%s", errMultipleBoxes, round, applicationID, encodedBoxName))
+	}
+
+	box := boxes[0]
+	if len(box.Name) == 0 && len(boxName) > 0 {
+		return notFound(ctx, fmt.Sprintf("%s: round=%d, appid=%d, boxName=%+v", errNoBoxesFound, round, applicationID, encodedBoxName))
+	}
+
+	if string(box.Name) != string(boxName) {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, boxName=%s", errWrongBoxFound, round, applicationID, encodedBoxName))
+	}
+
+	return ctx.JSON(http.StatusOK, generated.BoxResponse(box))
+}
+
+// SearchForApplicationBoxes returns box names for an app
+// (GET /v2/applications/{application-id}/boxes)
+func (si *ServerImplementation) SearchForApplicationBoxes(ctx echo.Context, applicationID uint64, params generated.SearchForApplicationBoxesParams) error {
+	if err := si.verifyHandler("SearchForApplicationBoxes", ctx); err != nil {
+		return badRequest(ctx, err.Error())
+	}
+
+	q := idb.ApplicationBoxQuery{
+		ApplicationID: applicationID,
+		OmitValues:    true,
+	}
+	if params.Limit != nil {
+		q.Limit = *params.Limit
+	}
+	if params.Next != nil {
+		encodedBoxName := *params.Next
+		boxNameBytes, err := logic.NewAppCallBytes(encodedBoxName)
+		if err != nil {
+			return badRequest(ctx, err.Error())
+		}
+		prevBox, err := boxNameBytes.Raw()
+		if err != nil {
+			return badRequest(ctx, err.Error())
+		}
+		q.PrevFinalBox = prevBox
+	}
+
+	appid, boxes, round, err := si.fetchApplicationBoxes(ctx.Request().Context(), q)
+
+	if err != nil {
+		msg := fmt.Sprintf("%s: round=?=%d, appid=%d", errFailedSearchingBoxes, round, applicationID)
+		return indexerError(ctx, fmt.Errorf("%s: %w", msg, err))
+	}
+
+	if len(boxes) == 0 {
+		return notFound(ctx, fmt.Sprintf("%s: round=%d, appid=%d", errNoApplicationsFound, round, applicationID))
+	}
+
+	if appid != generated.ApplicationId(applicationID) {
+		return indexerError(ctx, fmt.Errorf("%s: round=%d, appid=%d, wrong appid=%d", errWrongAppidFound, round, applicationID, appid))
+	}
+
+	var next *string
+	finalNameBytes := boxes[len(boxes)-1].Name
+	if finalNameBytes != nil {
```

Comment:
@tzaffi Request for clarification:  In what scenarios can `finalNameBytes == nil`?

From what I understand about upstream codepaths, my intuition is that the value  is always non-nil.

Reply:
For completeness, there's another `nil` check several lines below where I have the same question.

Reply:
`finalNameBytes == nil` in the case that the app-id is valid, but the query resulted in no boxes.

Regarding the second `nil` check where I'm skipping nil boxes, a similar effect could be had by having more explicit boundary checks. This approach -while lazy- seemed more robust against query variations. However, I'm happy to dig further with additional test cases and use more explicit checks instead of the current approach. LMK your thoughts.

Reply:
@tzaffi Resolving - Makes sense.  Thanks for the explanation and sorry for the silly question.

Due to my prior experiences, there are circumstances where my mind still operates differently.  It causes me to get tangled like I did here.